File: ruby/lib/hachi64.rb
# frozen_string_literal: true

# Hachi64: 哈吉米64 Encoding and Decoding
# This module provides encoding and decoding using the Hachi64 character set.

module Hachi64
  # The Hachi64 alphabet - 64 unique Chinese characters
  HACHI_ALPHABET = "哈蛤呵吉急集米咪迷南男难北背杯绿律虑豆斗抖啊阿额西希息嘎咖伽花华哗压鸭呀库酷苦奶乃耐龙隆拢曼慢漫波播玻叮丁订咚东冬囊路陆多都弥济"

  # Encodes a byte string into a Hachi64 string using the Hachi64 alphabet.
  #
  # @param data [String] The bytes to encode (binary string)
  # @param padding [Boolean] Whether to use '=' for padding (default: true)
  # @return [String] The encoded string
  def self.encode(data, padding: true)
    alphabet = HACHI_ALPHABET
    result = []
    data_len = data.bytesize
    i = 0

    while i < data_len
      chunk = data.byteslice(i, 3)
      i += 3

      byte1 = chunk.getbyte(0)
      byte2 = chunk.bytesize > 1 ? chunk.getbyte(1) : 0
      byte3 = chunk.bytesize > 2 ? chunk.getbyte(2) : 0

      idx1 = byte1 >> 2
      idx2 = ((byte1 & 0x03) << 4) | (byte2 >> 4)
      idx3 = ((byte2 & 0x0F) << 2) | (byte3 >> 6)
      idx4 = byte3 & 0x3F

      result << alphabet[idx1]
      result << alphabet[idx2]

      if chunk.bytesize > 1
        result << alphabet[idx3]
      elsif padding
        result << '='
      end

      if chunk.bytesize > 2
        result << alphabet[idx4]
      elsif padding
        result << '='
      end
    end

    result.join
  end

  # Decodes a Hachi64 string into bytes using the Hachi64 alphabet.
  #
  # @param encoded_str [String] The string to decode
  # @param padding [Boolean] Whether the input string uses '=' for padding (default: true)
  # @return [String] The decoded bytes (binary string)
  # @raise [ArgumentError] If the input string contains invalid characters
  def self.decode(encoded_str, padding: true)
    alphabet = HACHI_ALPHABET
    reverse_map = {}
    alphabet.each_char.with_index { |char, i| reverse_map[char] = i }

    # Handle empty string
    return ''.b if encoded_str.empty?

    pad_count = 0
    if padding
      pad_count = encoded_str.count('=')
      if pad_count > 0
        encoded_str = encoded_str[0...-pad_count]
      end
    end

    result = []
    data_len = encoded_str.length
    i = 0

    while i < data_len
      chunk_end = [i + 4, data_len].min
      chunk = encoded_str[i...chunk_end]
      i += 4

      idx1 = reverse_map[chunk[0]]
      idx2 = chunk.length > 1 ? reverse_map[chunk[1]] : 0
      idx3 = chunk.length > 2 ? reverse_map[chunk[2]] : 0
      idx4 = chunk.length > 3 ? reverse_map[chunk[3]] : 0

      raise ArgumentError, "Invalid character in input" if idx1.nil? || (chunk.length > 1 && idx2.nil?)

      byte1 = (idx1 << 2) | (idx2 >> 4)
      result << byte1

      if chunk.length > 2
        byte2 = ((idx2 & 0x0F) << 4) | (idx3 >> 2)
        result << byte2
      end

      if chunk.length > 3
        byte3 = ((idx3 & 0x03) << 6) | idx4
        result << byte3
      end
    end

    result.pack('C*')
  end
end

File: go/hachi64.go
// Package hachi64 provides encoding and decoding functions for Hachi64,
// a Base64-style encoding using 64 Chinese characters.
//
// 哈基米64 编解码器 - 使用64个中文字符进行 Base64 风格的编码和解码。
package hachi64

import (
	"fmt"
	"strings"
)

// HachiAlphabet 哈基米64字符集：64个中文字符，按同音字分组
const HachiAlphabet = "哈蛤呵吉急集米咪迷南男难北背杯绿律虑豆斗抖啊阿额西希息嘎咖伽花华哗压鸭呀库酷苦奶乃耐龙隆拢曼慢漫波播玻叮丁订咚东冬囊路陆多都弥济"

// alphabetRunes contains the Hachi alphabet as a slice of runes for indexing
var alphabetRunes []rune

// reverseMap maps each character to its index in the alphabet
var reverseMap map[rune]int

func init() {
	alphabetRunes = []rune(HachiAlphabet)
	if len(alphabetRunes) != 64 {
		panic("HachiAlphabet must contain exactly 64 characters")
	}

	reverseMap = make(map[rune]int, 64)
	for i, r := range alphabetRunes {
		reverseMap[r] = i
	}
}

// Encode encodes the given byte slice into a Hachi64 string.
//
// Parameters:
//   - data: The bytes to encode
//   - padding: Whether to use '=' for padding (default true)
//
// Returns the encoded string.
func Encode(data []byte, padding bool) string {
	if len(data) == 0 {
		return ""
	}

	var result strings.Builder
	result.Grow((len(data)/3 + 1) * 4)

	// Process data in chunks of 3 bytes
	for i := 0; i < len(data); i += 3 {
		// Get byte values, padding with 0 if needed
		byte1 := data[i]
		var byte2, byte3 byte
		chunkLen := 1

		if i+1 < len(data) {
			byte2 = data[i+1]
			chunkLen = 2
		}
		if i+2 < len(data) {
			byte3 = data[i+2]
			chunkLen = 3
		}

		// Split 24 bits into 4 6-bit indices
		idx1 := byte1 >> 2
		idx2 := ((byte1 & 0x03) << 4) | (byte2 >> 4)
		idx3 := ((byte2 & 0x0F) << 2) | (byte3 >> 6)
		idx4 := byte3 & 0x3F

		// Add first two characters (always present)
		result.WriteRune(alphabetRunes[idx1])
		result.WriteRune(alphabetRunes[idx2])

		// Handle third character
		if chunkLen > 1 {
			result.WriteRune(alphabetRunes[idx3])
		} else if padding {
			result.WriteRune('=')
		}

		// Handle fourth character
		if chunkLen > 2 {
			result.WriteRune(alphabetRunes[idx4])
		} else if padding {
			result.WriteRune('=')
		}
	}

	return result.String()
}

// Decode decodes a Hachi64 string back into bytes.
//
// Parameters:
//   - encodedStr: The string to decode
//   - padding: Whether the input string uses '=' for padding (default true)
//
// Returns the decoded bytes and any error encountered.
func Decode(encodedStr string, padding bool) ([]byte, error) {
	if encodedStr == "" {
		return []byte{}, nil
	}

	// Handle padding
	padCount := 0
	if padding {
		padCount = strings.Count(encodedStr, "=")
		if padCount > 0 {
			encodedStr = strings.TrimRight(encodedStr, "=")
		}
	}

	runes := []rune(encodedStr)
	result := make([]byte, 0, (len(runes)*3)/4)

	// Process in chunks of 4 characters
	for i := 0; i < len(runes); i += 4 {
		chunkLen := 4
		if i+4 > len(runes) {
			chunkLen = len(runes) - i
		}

		// Get indices
		idx1, ok := reverseMap[runes[i]]
		if !ok {
			return nil, fmt.Errorf("invalid character in input: %c", runes[i])
		}

		var idx2, idx3, idx4 int
		if chunkLen > 1 {
			idx2, ok = reverseMap[runes[i+1]]
			if !ok {
				return nil, fmt.Errorf("invalid character in input: %c", runes[i+1])
			}
		}
		if chunkLen > 2 {
			idx3, ok = reverseMap[runes[i+2]]
			if !ok {
				return nil, fmt.Errorf("invalid character in input: %c", runes[i+2])
			}
		}
		if chunkLen > 3 {
			idx4, ok = reverseMap[runes[i+3]]
			if !ok {
				return nil, fmt.Errorf("invalid character in input: %c", runes[i+3])
			}
		}

		// Reconstruct 3 bytes from 4 6-bit indices
		byte1 := byte((idx1 << 2) | (idx2 >> 4))
		result = append(result, byte1)

		if chunkLen > 2 {
			byte2 := byte(((idx2 & 0x0F) << 4) | (idx3 >> 2))
			result = append(result, byte2)
		}

		if chunkLen > 3 {
			byte3 := byte(((idx3 & 0x03) << 6) | idx4)
			result = append(result, byte3)
		}
	}

	return result, nil
}
